### PR TITLE
consensus: handle justification sync for blocks authored locally

### LINKED ISSUE #4678: Invalid justification provided
Sometimes, the nodes report `Invalid justification provided` then it will fall behind the latest node few blocks. And I've to restart those node to (re)participate in validating.

### PR DESCRIPTION
**Pull request by andresilva**
When importing new blocks different consensus engines might require that finality data is provided for that block (i.e. a justification), or it might require that existing requests for justifications are dropped. This is signalled as part of the `ImportAux` object that is part of `ImportResult::Imported` and is handled by the import queue, i.e. after importing blocks the import queue will check `ImportAux` and issue justification requests to the sync module. 

When authoring blocks (through slot worker or pow miner) the blocks are imported directly without going through the import queue, in which case any required justification handling was ignored. This PR makes all existing block authoring handle justification requirements similarly to the import queue.

Fix #4678.

polkadot companion: https://github.com/paritytech/polkadot/pull/2956

### CODE REVIEW

**Review by octol — APPROVED** [review]

File: bin/node-template/node/src/service.rs [hunk 1]
```diff
@@ -230,7 +230,7 @@ pub fn new_full(mut config: Configuration) -> Result<TaskManager, ServiceError>
 		let slot_duration = sc_consensus_aura::slot_duration(&*client)?;
 		let raw_slot_duration = slot_duration.slot_duration();
 
-		let aura = sc_consensus_aura::start_aura::<AuraPair, _, _, _, _, _, _, _, _, _, _>(
+		let aura = sc_consensus_aura::start_aura::<AuraPair, _, _, _, _, _, _, _, _, _, _, _>(
```
Comment (bkchr) [inline bin/node-template/node/src/service.rs:233]: :see_no_evil:

File: client/consensus/slots/src/lib.rs [hunk 2]
```diff
@@ -77,6 +79,7 @@ pub trait SlotWorker<B: BlockT, Proof> {
 	async fn on_slot(
 		&mut self,
 		slot_info: SlotInfo<B>,
+		justification_sync_link: &mut dyn JustificationSyncLink<B>,
```
Comment (bkchr) [inline client/consensus/slots/src/lib.rs:82]: Please move this to the `SimpleSlotWorker` trait as it is done with the other values we need in the `on_slot` implementation.,

File: client/consensus/babe/src/lib.rs [hunk 3]
```diff
@@ -399,6 +399,9 @@ pub struct BabeParams<B: BlockT, C, E, I, SO, SC, CAW, BS, IDP> {
 	/// Checks if the current native implementation can author with a runtime at a given block.
 	pub can_author_with: CAW,
 
+	/// Hook into the sync module to control the justification sync process.
+	pub justification_sync_link: L,
```
Comment (bkchr) [inline client/consensus/babe/src/lib.rs:403]: We should really rewrite all this stuff to have *one* block import and not this stuff here that is split :see_no_evil:

File: client/consensus/slots/src/lib.rs (hunk 2 above)
Comment (andresilva) [reply client/consensus/slots/src/lib.rs:82]: Yeah I was a bit lazy, also the reason I'm using a trait object here 😊. Will update.

File: client/consensus/babe/src/lib.rs (hunk 3 above)
Comment (andresilva) [reply client/consensus/babe/src/lib.rs:403]: Do you mean having the exact same process for importing locally authored and synced blocks? If that's you mean then I agree as I also don't like having to duplicate this logic from the import queue, but that requires a bigger refactoring 🥲.

File: bin/node-template/node/src/service.rs (hunk 1 above)
Comment (andresilva) [reply bin/node-template/node/src/service.rs:233]: All because keypair type can't be inferred. Life's hard.

**Review by bkchr — APPROVED** [review]

**Comment by andresilva** [comment]
bot merge

**Pull request merged**